### PR TITLE
Refactor BlockInProgress to print more information of processed blocks and chunks

### PR DESCRIPTION
**Pull request by mzhangmzz**
This PR refactors BlockInProgress to keep information of recent blocks, not just blocks that are in progress and updates the debug page and debug stats print. 

The new implementation also fixed some small issues: 
- Implemented GC in BlockInProgress correctly. Previously, a block is only removed if it finished processing. This could leave blocks permanently left in the pool if it never finished processing, which could happen for blocks on some old forks.
- Now missing chunks are reflected with character "X"

In a following PR, I will add more detailed information regarding chunk processing (which parts/receipts are received/missing).

New debug page: 
Screen Shot 2022-07-29 at 4 34 34 PM

New debug stats:
```
2022-07-29T19:00:06.422803Z DEBUG stats: EpochId(`AcpLCWttsHYuh7HNMoKaQxU63enpEJC31TXDZnFXDqVS`) Orphans: 0 With missing chunks: 0 In processing 0
1881 GdeK8aYv6JYmc2zriRZjYNxiPy3r5wBNzXpaN73d2cw7 Processed in progress for 16ms   Chunks:(✔✔✔✔))
1880 AeXssP35WTDmxLSRJBs7AP5KsBgbPWmLr5217JDSTCqy Processed in progress for 27ms   Chunks:(✔✔✔✔))
1879 Bj2eheEAoKeEUUM3mm6iHbc48cLfFSvnsGNsmmALzLyK Processed in progress for 11ms   Chunks:(✔✔✔✔))
1878 3qMZnZNSaWhRtCoBXkkFcVnwm1oLZcKPG2kfJ3HMtzhk Processed in progress for 19ms   Chunks:(✔✔✔✔))
1877 4pVuKs8PK3HTQXxh5iHHTnYphpVUyXyx135FJaTXwP71 Processed in progress for 9ms   Chunks:(✔✔✔✔))
1876 3QzRZHhb7dbWzbk3akLXzBppyuLtWTwVK9x9ytpXb9Mw Processed in progress for 15ms   Chunks:(✔✔✔✔))
1875 BS3vJJGwA9KDmknFMetMqdMnw8G5rUFurowdndjSjiBx Processed in progress for 16ms   Chunks:(✔✔✔✔))
1874 7uKcdBdzgUfdhQ7AADZa1ryPoTZXWhQJjFfSYuR9U3hP Processed in progress for 24ms   Chunks:(✔✔✔✔))
1873 9APUhnFMfcetU3ZGbjJLTfZhspj9fATJKUo19uNwvDRr Processed in progress for 17ms   Chunks:(✔✔✔✔))
1872 APvWDvtEb8dCJiniQHPRcpAXYngWCv1PnyVdGqfT63co Processed in progress for 17ms   Chunks:(✔✔✔✔))
1871 HZJUwENk3PFYY7z5bVM1c722thUMGN4UpPAkMuDDaK9P Processed in progress for 11ms   Chunks:(✔✔✔✔))
1870 9KMtaFBZLPyWrYxpR5zWVQXMw5WragYxkD9E8uryjvag Processed in progress for 14ms   Chunks:(✔✔✔✔))
1869 eoxu78RPr7fb64CynxAqWSHgR774bWivaFRX6ti8arg Processed in progress for 17ms   Chunks:(XXXX))
```

### CODE REVIEW

File: chain/chain/src/blocks_delay_tracker.rs [hunk 1]
```diff
@@ -1,29 +1,48 @@
...
-    pub blocks_in_progress: HashMap<CryptoHash, BlockInProgress>,
-    /// An entry gets created when a chunk gets requested for the first time.
-    /// Chunks get deleted when the block gets processed.
-    pub chunks_in_progress: HashMap<ChunkHash, ChunkInProgress>,
+    // A block is added at the first time it was received, and
+    // removed if it is too far from the chain head.
+    blocks: HashMap<CryptoHash, BlockTrackingStats>,
+    // Maps block height to block hash. Used for gc.
+    // Theoretically, each block height should only have one block, if our work works correctly.
+    blocks_height_map: BTreeMap<BlockHeight, Vec<CryptoHash>>, // Chunks that belong to the blocks in the tracker
```
Comment (mm-near) [inline chain/chain/src/blocks_delay_tracker.rs:36]: should this comment about chunks be in the next line?

File: chain/chain/src/blocks_delay_tracker.rs [hunk 2]
```diff
@@ -159,22 +287,218 @@ impl BlocksDelayTracker {
...
+
+                let in_progress_str = format!("in progress for {:?}ms", block_info.in_progress_ms);
+
+                let in_orphan_str = match block_info.orphaned_ms {
+                    Some(duration) => format!("orphan for {:?}ms", duration),
+                    None => "".to_string(),
+                };
+
+                let missing_chunks_str = match block_info.missing_chunks_ms {
+                    Some(duration) => format!("orphan for {:?}ms", duration),
```
Comment (mm-near) [inline chain/chain/src/blocks_delay_tracker.rs:477]: shouldn't this say 'missing chunks'?

File: core/primitives/src/views.rs [hunk 3]
```diff
@@ -383,11 +383,73 @@ pub struct BlockByChunksView {
...
+pub struct ChunkProcessingInfo {
+    pub height_created: BlockHeight,
+    pub shard_id: ShardId,
+    pub chunk_hash: ChunkHash,
+    pub prev_block_hash: CryptoHash,
+    /// Account id of the validator who created this chunk
+    pub created_by: Option<AccountId>,
+    pub status: ChunkProcessingStatus,
+    /*
+    /// Timestamp of first time when we request for this chunk.
```
Comment (mm-near) [inline core/primitives/src/views.rs:440]: remove?

File: core/primitives/src/views.rs [hunk 4]
```diff
@@ -383,11 +383,73 @@ pub struct BlockByChunksView {
...
+    pub floating_chunks_info: Vec<ChunkProcessingInfo>,
+}
+
+#[cfg_attr(feature = "deepsize_feature", derive(deepsize::DeepSizeOf))]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct BlockProcessingInfo {
+    pub height: BlockHeight,
+    pub hash: CryptoHash,
+    /// Timestamp when block was received.
+    //pub received_timestamp: DateTime<chrono::Utc>,
```
Comment (mm-near) [inline core/primitives/src/views.rs:402]: remove?

File: core/primitives/src/views.rs [hunk 5]
```diff
@@ -383,11 +383,73 @@ pub struct BlockByChunksView {
...
+}
+
+#[cfg_attr(feature = "deepsize_feature", derive(deepsize::DeepSizeOf))]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ChunkProcessingInfo {
+    pub height_created: BlockHeight,
+    pub shard_id: ShardId,
+    pub chunk_hash: ChunkHash,
+    pub prev_block_hash: CryptoHash,
+    /// Account id of the validator who created this chunk
```
Comment (mm-near) [inline core/primitives/src/views.rs:436]: also mention when this field can be None

**Review by mm-near — APPROVED** [review]

**Review by nikurt — APPROVED** [review]
LGTM

File: core/primitives/src/views.rs (hunk 4 above)
Comment (mzhangmzz) [reply core/primitives/src/views.rs:402]: same as last comment

**Pull request merged**